### PR TITLE
server: add a SendRunner to send raft messages to other stores.

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: src/server/raft_send.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::sync::Arc;
+use std::net::SocketAddr;
+
+use futures::sync::mpsc;
```
Comment (siddontang) [inline src/server/raft_send.rs]: `
use futures::sync::mpsc::{self, UnboundedSender};
`

File: src/server/raft_send.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+struct Conn {
+    _client: TikvClient,
+    stream: mpsc::UnboundedSender<RaftMessage>,
+}
+
+impl Conn {
+    fn new(env: Arc<Environment>, addr: SocketAddr, handle: &Handle) -> Conn {
+        let channel = ChannelBuilder::new(env).connect(&format!("{}", addr));
```
Comment (siddontang) [inline src/server/raft_send.rs]: suggest passing an Options, not only SocketAddr, to support more Builder args later.

File: src/server/raft_send.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    fn get_conn(&mut self, addr: SocketAddr, handle: &Handle) -> &Conn {
+        let env = self.env.clone();
+        self.conns
+            .entry(addr)
+            .or_insert_with(|| Conn::new(env.clone(), addr, handle))
+    }
+
+    fn send(&mut self, t: SendTask, handle: &Handle) -> Result<()> {
+        let conn = self.get_conn(t.addr, handle);
+        box_try!(mpsc::UnboundedSender::send(&conn.stream, t.msg));
```
Comment (siddontang) [inline src/server/raft_send.rs]: can we use `conn.stream.send` directly?

File: src/server/raft_send.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+use kvproto::tikvpb_grpc::TikvClient;
+
+use util::collections::HashMap;
+use util::worker::FutureRunnable;
+use super::{Error, Result};
+
+/// SendTask delivers a raft message to other store.
+pub struct SendTask {
+    pub addr: SocketAddr,
+    pub msg: RaftMessage,
```
Comment (siddontang) [inline src/server/raft_send.rs]: maybe we can support sending batch messages later.

File: src/server/raft_send.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+impl Conn {
+    fn new(env: Arc<Environment>, addr: SocketAddr, handle: &Handle) -> Conn {
+        let channel = ChannelBuilder::new(env).connect(&format!("{}", addr));
+        let client = TikvClient::new(channel);
+        let (tx, rx) = mpsc::unbounded();
+        handle.spawn(client.raft()
+            .sink_map_err(Error::from)
+            .send_all(rx.map_err(|_| Error::Sink))
+            .map(|_| ())
+            .map_err(|e| error!("send raftmessage failed: {:?}", e)));
```
Comment (siddontang) [inline src/server/raft_send.rs]: can we know the address in error log? so that we can know which store is unreachable.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/server/raft_send.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            env: env,
+            conns: HashMap::default(),
+        }
+    }
+
+    fn get_conn(&mut self, addr: SocketAddr, handle: &Handle) -> &Conn {
+        let env = self.env.clone();
+        self.conns
+            .entry(addr)
+            .or_insert_with(|| Conn::new(env.clone(), addr, handle))
```
Comment (overvenus) [inline src/server/raft_send.rs]: How about`|| Conn::new(self.env.clone(), addr, handle)`?

File: src/server/raft_send.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+impl Conn {
+    fn new(env: Arc<Environment>, addr: SocketAddr, handle: &Handle) -> Conn {
+        let channel = ChannelBuilder::new(env).connect(&format!("{}", addr));
+        let client = TikvClient::new(channel);
+        let (tx, rx) = mpsc::unbounded();
+        handle.spawn(client.raft()
+            .sink_map_err(Error::from)
+            .send_all(rx.map_err(|_| Error::Sink))
```
Comment (overvenus) [inline src/server/raft_send.rs]: `send_all` waits until a stream ends, how about: 
```
let s = client.sink().sink_map_err(Error::from);
rx.for_each(|i| {
    s.send(i);
    Ok(())
});
```

File: src/server/raft_send.rs (hunk 2 above)
Comment (disksing) [inline src/server/raft_send.rs]: I think it won't be too late to refactor it when we really need other options.

File: src/server/raft_send.rs (hunk 7 above)
Comment (disksing) [inline src/server/raft_send.rs]: It won't compile:
```
error[E0507]: cannot move out of captured outer variable in an `FnMut` closure
  --> src/server/raft_send.rs:53:13
   |
53 |             s.send(i);
   |             ^ cannot move out of captured outer variable in an `FnMut` closure
```

File: src/server/raft_send.rs (hunk 3 above)
Comment (disksing) [inline src/server/raft_send.rs]: The compiler says:
```
error[E0507]: cannot move out of borrowed content
  --> src/server/raft_send.rs:86:9
   |
86 |         conn.stream.send(t.msg);
   |         ^^^^ cannot move out of borrowed content
```

Comment (siddontang) [inline src/server/raft_send.rs]: 😲

File: src/server/raft_send.rs (hunk 7 above)
Comment (siddontang) [inline src/server/raft_send.rs]: @BusyJay 

Any idea?

Comment (BusyJay) [inline src/server/raft_send.rs]: I don't get it, what's wrong with `send_all`?

Comment (siddontang) [inline src/server/raft_send.rs]: /cc @overvenus

Comment (overvenus) [inline src/server/raft_send.rs]: It's fine, my mistake. :-p

**Comment by siddontang** [comment]
LGTM

**Pull request merged**